### PR TITLE
Make sure built-in procedures close resources when done

### PR DESCRIPTION
**Pull request by systay**
The first two commits are backports from 3.3 and should be forward-merged with a null commit. The last commit is applicable to both 3.2 and 3.3

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInProcedures.java [hunk 1]
```diff
@@ -67,17 +67,20 @@
...
     @Description( "List all property keys in the database." )
     @Procedure( name = "db.propertyKeys", mode = READ )
     public Stream<PropertyKeyResult> listPropertyKeys()
     {
+
         try ( Statement statement = tx.acquireStatement() )
         {
-            return TokenAccess.PROPERTY_KEYS.inUse( statement ).map( PropertyKeyResult::new ).stream();
+            // Ownership of the reference to the acquired statement is transfered to the returned iterator stream
+            return asList( TokenAccess.PROPERTY_KEYS.inUse( statement ).map( PropertyKeyResult::new ) ).stream();
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInProcedures.java]: Yup, this eagerness here is the preferred way to go

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/TokenAccess.java [hunk 2]
```diff
@@ -143,7 +143,11 @@ private static boolean hasAny( Iterator<?> iter )
         @Override
         public void close()
         {
-            statement.close();
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/TokenAccess.java:146]: Is this strictly necessary or just being over the top safe?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ReflectiveProcedureCompiler.java [hunk 3]
```diff
@@ -671,7 +671,11 @@ public boolean hasNext() throws ProcedureException
             @Override
             public void close() throws Exception
             {
-                closeable.close();
+                if ( closeable != null )
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ReflectiveProcedureCompiler.java:674]: Is this strictly necessary or just being over the top safe?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/TokenAccess.java (hunk 2 above)
Comment (systay) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/api/TokenAccess.java:146]: This was back-ported from the 3.3 branch. Not sure - what do you think, @henriknyman?

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/ProcedureCallSupportAcceptanceTest.scala [hunk 4]
```diff
@@ -87,4 +87,26 @@ class ProcedureCallSupportAcceptanceTest extends ProcedureCallAcceptanceTest {
       map("one", "alpha", "newTwo", "beta")
     ))
   }
+
+  test("should be able to execute union of multiple token accessing procedures") {
```
Comment (thobe) [inline enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/ProcedureCallSupportAcceptanceTest.scala:91]: This test fails with:

```
org.neo4j.graphdb.NotInTransactionException: The statement has been closed.
```

**Comment by thobe** [comment]
Just to verify my understanding of this changeset. What is supposed to fix the issue is mainly:

- [Line 102 in DefaultExecutionResultBuilderFactory](https://github.com/neo4j/neo4j/pull/10071/files#diff-9e953639cb3fbc71d1fa8d742723cce0R102)
- [Line 102 in ProcedureExecutionResult](https://github.com/neo4j/neo4j/pull/10071/files#diff-1257d865b16a46ad191c188d40ca02ccR102), and
- [Line 54 in CompiledExecutionResult](https://github.com/neo4j/neo4j/pull/10071/files#diff-123545a8c821d04dbcd74857dc4c0545R54)

Then there are changes to some of the procedures to ensure that they finish their reads before the statement is closed.

Is that correct?

**Comment by systay** [comment]
We are trying to make the test `should be able to execute union of multiple token accessing procedures` go green, and those are the changes needed.

**Comment by henriknyman** [comment]
@thobe The three things you mention are a secondary issue when we run in PROFILE mode. The main issue is that the three built-in procedures could not close the statement while returning a lazy stream.

**Comment by henriknyman** [comment]
@tinwelint The reason we have to transfer ownership of the statement into `TokenAccess` is because that is the way the Core API works. The statement will be closed when the iterator/stream is exhausted.
We could make the built-in procedures eager if we are worried about leaking a statement in case of an exception, but we cannot do try-with-resource since the statement will be closed by exhausting the stream in the happy case.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/TokenAccess.java (hunk 2 above)
Comment (henriknyman) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/api/TokenAccess.java:146]: Hmm I don't remember.. it could be the latter.

**Comment by tinwelint** [comment]
@henriknyman OK right, so that will only get closed in the happy case (no exception and stream fully consumed) if letting the stream manage the closing of the resource. This eagerness feels better and is typically on very small number of tokens.

**Review by tinwelint — APPROVED** [review]

**Comment by thobe** [comment]
@systay / @henriknyman: this is still not where it needs to be:

```
java.lang.AssertionError: Failed on procedure CALL db.propertyKeys()
    at org.neo4j.kernel.builtinprocs.ProcedureResourcesIT.verifyProcedureCloseAllAcquiredKernelStatements(ProcedureResourcesIT.java:118)
    at org.neo4j.kernel.builtinprocs.ProcedureResourcesIT.allProcedures(ProcedureResourcesIT.java:82)
```

**Pull request merged**